keps/sig-auth/2799-reduction-of-secret-based-service-account-token/README.md

```diff
@@ -0,0 +1,463 @@
+  secrets might stop working. To mitigate,
+  1. Emit warnings when using auto-generated token secrets.
+  2. Publish pointers to TokenRequest or the manual secret request flow.
+- When LegacyServiceAccountTokenCleanUp is Beta, usage of auto-generated
+  secret-based token might stop working. To mitigate,
+  1. When Alpha, annouce the cleanup starts at Beta
+  2. Emit warnings when using auto-generated token secrets.
+  3. Add pointers of TokenRequest API and manually created tokens in the validation
+     result.
+
+## Design Details
+
+### LegacyServiceAccountTokenNoAutoGeneration:
+
+Token Controller stops auto-creating secret for service accounts. This feature would
+be enabled when it is implemented since no new code is added and this can make
+sure new clusters are in good state.
+
+### LegacyServiceAccountTokenTracking
+
+To facilitate LegacyServiceAccountTokenCleanUp, we implement a simple controller
+in kube-apiserver that maintains a bool value configmap in `kube-system` to
+indicates if tracking is enabled in the cluster. It is similar to the existing
+`ClusterAuthenticationTrustController` that maintains `configmap/extension-apiserver-authentication`.
+
+- When LegacyServiceAccountTokenTracking is enabled in all apiservers,
+
+  - the controller creates/updates a configmap in `kube-system` namespace that
+    stores the current date as `tracked-since`.
+  - when a legacy token is used, issue a warning, annotate/update the
+    `last-used` on the secret, and record in a metric. optionally, add a
+    label `in-use` for fast query.
+
+- When LegacyServiceAccountTokenTracking is disabled in any apiserver,
+  - the controller ensures the configmap in `kube-system` namespace is deleted
+    in a periodic way.
+
+### LegacyServiceAccountTokenCleanUp
+
+Token Controller starts to remove unused auto-generated secrets (secrets
+bi-directionally referenced by the service account) and not mounted by pods.
+
+When this feature is Beta and enabled by default, delete secrets iff it is over
+a sufficient period of time (one year by default) since last used. The period
+can be configured by cluster admins.
+
+Determine the date that a given secret was last used:
+
+1. `last-used` if exists and after `tracked-since`.
+2. defaults to `tracked-since`
+
+### Test Plan
+
+- Unit tests
+- Integration tests
+  - Previously auto-generated secret-based token that's used within the
+    configurable cleanup duration will continue to work.
+  - Previously auto-generated secret-based token that's used after the
+    configurable cleanup duration will be deleted.
+- E2E tests
+  - Secret-based tokens would not be auto-generated.
+  - Still able to explicitly request a secret-based token.
+  - The explicitly requested token would not be deleted.
+
+### Graduation Criteria
+
+#### LegacyServiceAccountTokenNoAutoGeneration
+
+| Alpha | Beta | GA   |
+| ----- | ---- | ---- |
+| -     | 1.24 | 1.25 |
```

Comment:
Can you describe why this starts as beta?

Reply:
because in 1.24, all pods should be admitted in 1.22+ and they should be using bound tokens. this feature gate mostly guard deletions of code.

Reply:
> because in 1.24, all pods should be admitted in 1.22+ and they should be using bound tokens. this feature gate mostly guard deletions of code.

I meant in the KEP itself. It's unusual to start featuregates in beta.  Not unheard of, but worthy of explaining in the doc why it's more reasonable to start this one at beta than at alpha.  Why should is it more reasonable to start this as opt-out than as opt-in?

Reply:
updated.